test_conformance/printf/util_printf.cpp

```diff
@@ -859,16 +786,130 @@ size_t verifyOutputBuffer(char *analysisBuffer,testCase* pTestCase,size_t testId
     return strcmp(analysisBuffer,"inf")&&strcmp(analysisBuffer,"infinity")&&strcmp(analysisBuffer,"1.#INF00")&&strcmp(analysisBuffer,"Inf");
-    if(!strcmp(pTestCase->_correctBuffer[testId],"nan") || !strcmp(pTestCase->_correctBuffer[testId],"-nan")) {
+    if(!strcmp(pTestCase->_correctBuffer[testId].c_str(),"nan") || !strcmp(pTestCase->_correctBuffer[testId].c_str(),"-nan")) {
        return strcmp(analysisBuffer,"nan")&&strcmp(analysisBuffer,"-nan")&&strcmp(analysisBuffer,"1.#IND00")&&strcmp(analysisBuffer,"-1.#IND00")&&strcmp(analysisBuffer,"NaN")&&strcmp(analysisBuffer,"nan(ind)")&&strcmp(analysisBuffer,"nan(snan)")&&strcmp(analysisBuffer,"-nan(ind)");
     }
-    return strcmp(analysisBuffer,pTestCase->_correctBuffer[testId]);
+    return strcmp(analysisBuffer,pTestCase->_correctBuffer[testId].c_str());
+}
+
+static void intRefBuilder(printDataGenParameters& params)
+{
+    printf(params.genericFormat, atoi(params.dataRepresentation));
+}
+
+static void floatRefBuilder(printDataGenParameters& params)
+{
+    printf(params.genericFormat, strtof(params.dataRepresentation, NULL));
+}
+
+static void octalRefBuilder(printDataGenParameters& params)
+{
+    const unsigned long int data = strtoul(params.dataRepresentation, NULL, 10);
+    printf(params.genericFormat, data);
+}
+
+static void unsignedRefBuilder(printDataGenParameters& params)
+{
+    const unsigned long int data = strtoul(params.dataRepresentation, NULL, 10);
+    printf(params.genericFormat, data);
+}
+
+static void hexRefBuilder(printDataGenParameters& params)
+{
+    const unsigned long int data = strtoul(params.dataRepresentation, NULL, 0);
+    printf(params.genericFormat, data);
 }
 
+/*
+    Generate reference results.
+
+    Results are only generated for test cases
+    that can easily be generated by using CPU
+    printf.
+
+    If that is not the case, results are constants
+    that have been hard-coded.
+*/
+void generateRef(const cl_device_id device)
+{
+    int fd = -1;
+    char _refBuffer[ANALYSIS_BUFFER_SIZE];
+    cl_device_fp_config fpConfig;
+    const RoundingMode hostRound = get_round();
+    RoundingMode deviceRound;
+
+    // Discover device rounding mode
+    cl_int err = clGetDeviceInfo(device,
+                                CL_DEVICE_SINGLE_FP_CONFIG,
+                                sizeof(fpConfig),
+                                &fpConfig,
+                                NULL);
+    if (err != CL_SUCCESS)
+    {
+        log_error("clGetDeviceInfo failed");
+        return;
+    }
+
+    // Map device rounding to CTS rounding type
+    if (fpConfig & CL_FP_ROUND_TO_NEAREST)
+    {
+        deviceRound = kRoundToNearestEven;
+    }
+    else if (fpConfig & CL_FP_ROUND_TO_ZERO)
+    {
+        deviceRound = kRoundTowardZero;
+    }
+    else if (fpConfig & CL_FP_ROUND_TO_INF)
+    {
+        // Can be either round up or round down
+        deviceRound = kRoundUp;
+    }
+    else
+    {
+        assert(true && "Unreachable");
```

Comment:
I'm guessing you meant `assert(false && "Unreachable")`

Reply:
Good spot, yeah this should be false. I'll update it.